pkg/configuration/storage.go

```diff
@@ -107,3 +127,28 @@ func (s *JsonStorage) Set(key string, value any) error {
 
 	return err
 }
+
+func (s *JsonStorage) Refresh(config Configuration) error {
+	contents, err := os.ReadFile(s.path)
+	if err != nil {
+		return err
+	}
+	doc := map[string]interface{}{}
+	err = json.Unmarshal(contents, &doc)
+	if err != nil {
+		return err
+	}
+	for k, v := range doc {
+		config.Set(k, v)
```

Comment:
Possible Issue: Using Set() is an override, and overrides have precedence. This means that setting values like this might cause a change in behaviour. Can we make use of the reloading mechanism inside viper, for example by calling [ReadInConfig()](https://pkg.go.dev/github.com/spf13/viper#ReadInConfig) instead?

Reply:
In trying this approach, I think I've found a pre-existing bug in GAF configuration. `ev.viper.ReadInConfig()` fails to update the extendedViper keys in my unit tests. Digging into it, here's what I found:

The JsonStorage created by extendedViper always writes `~/.config/configstore/snyk.json` -- this filename is hard-coded, regardless of what the config name is set to, see https://github.com/snyk/go-application-framework/blob/068ee27b9941a958cce67e11d7eb332302678aaa/pkg/configuration/configuration.go#L465

However ReadInConfig attempts to honor the viper config name ("test" in unit tests), fails to find the file, and fails to refresh as a result. This might not impact production code if we only use `snyk.json` there, but it makes me aware that our storage implementation isn't always consistent with the way we're orchestrating viper.

I'm concerned this is becoming a rabbit hole and something to fix in a followup. For the time being, points to consider:

1. `Set` is used by `Clone`, so the override might be a reasonable thing to use in `Refresh`, given that we're setting from raw JSON parsed values.

2. Regarding keys not in the config, given that we persist only some of the keys, leaving the un-persisted config keys intact (as the current implementation does) seems less disruptive.

3. Making Storage entirely responsible for the serialization in refresh will likely have better results for a LS alternative implementation.

Reply:
Trying to separate my concern above from a solution proposal.

**Concern**
Using Set() for all values from the config file changes the precedence. Taking the example where `org` is set in the config file but also as an Environment Variable. The normal behaviour is that the environment variable will be consumed. After udpating the token, the org value from the config file will be Set() and suddenly another org will be used.

I think this concern is generally valid here. The difference to the mentioned usage in `Clone()` is that there it calls `Get()` and uses this value to `Set()`, it basically takes a snapshot of the configuration in time. But the snapshot has the same values. And doesn't change automatically through the update of an unassociated value.

**Solution**
If my previous hint towards a solution doesn't work, let's take a look further.